File: Baekjoon/5582.java
import java.io.*;

class Main {
    static int[][] dp = new int[4001][4001];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String a = br.readLine();
        String b = br.readLine();

        int ans = 0;
        for (int i = 1; i <= a.length(); i++) {
            for (int j = 1; j <= b.length(); j++) {
                if (a.charAt(i - 1) == b.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                    ans = Integer.max(ans, dp[i][j]);
                }
            }
        }
        bw.write(ans + "\n");
        bw.flush();
    }
}

File: Baekjoon/5582.kt
import java.lang.Integer.max

fun main() {
    val dp = Array(4001) { IntArray(4001) { 0 } }

    val a = readLine()!!
    val b = readLine()!!
    var ans = 0
    for (i in 1..a.length) {
        for (j in 1..b.length) {
            if (a[i - 1] == b[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1
                ans = max(ans, dp[i][j])
            }
        }
    }
    println(ans)
}
